connector/profiler/src/main/scala/org/apache/spark/executor/profiler/ExecutorJVMProfiler.scala

```diff
@@ -0,0 +1,183 @@
+        logInfo("JVM profiler stopped")
+        running = false
+        finishWriting()
+      })
+    }
+  }
+
+  private def startWriting(): Unit = {
+    if (profilerDfsDir.isDefined) {
+      val applicationId = try {
+        conf.getAppId
+      } catch {
+        case _: NoSuchElementException => "local-" + System.currentTimeMillis
+      }
+      val config = SparkHadoopUtil.get.newConfiguration(conf)
+      val appName = conf.get("spark.app.name").replace(" ", "-")
+      val profilerOutputDirname = profilerDfsDir.get
+
+      val profileOutputFile =
+        s"$profilerOutputDirname/$applicationId/profile-$appName-exec-$executorId.jfr"
+      val fs = FileSystem.get(new URI(profileOutputFile), config);
+      val filenamePath = new Path(profileOutputFile)
+      outputStream = fs.create(filenamePath)
+      try {
+        if (fs.exists(filenamePath)) {
+          fs.delete(filenamePath, true)
+        }
+        logInfo(s"Copying executor profiling file to $profileOutputFile")
+        inputStream = new BufferedInputStream(new FileInputStream(s"$profilerLocalDir/profile.jfr"))
+        threadpool = ThreadUtils.newDaemonSingleThreadScheduledExecutor("profilerOutputThread")
+        threadpool.scheduleWithFixedDelay(new Runnable() {
+          override def run(): Unit = writeChunk(false)
+        }, writeInterval, writeInterval,
+          TimeUnit.SECONDS)
+        writing = true
+      } catch {
+        case e: Exception =>
+          logError("Failed to start JVM profiler", e)
+          if (threadpool != null) {
+            threadpool.shutdownNow()
+          }
+          if (inputStream != null) {
+            inputStream.close()
+          }
+          if (outputStream != null) {
+            outputStream.close()
+          }
+      }
+    }
+  }
+
+  private def writeChunk(lastChunk: Boolean): Unit = {
+    if (!writing) {
+      return
+    }
+    try {
+      // stop (pause) the profiler, dump the results and then resume. This is not ideal as we miss
+      // the events while the file is being dumped, but that is the only way to make sure that
+      // the chunk of data we are copying to dfs is in a consistent state.
+      profiler.get.execute(stopcmd)
+      profiler.get.execute(dumpcmd)
+      var remaining = inputStream.available()
+      if (!lastChunk) {
+        profiler.get.execute(resumecmd)
+      }
+      while (remaining > 0) {
+        val read = inputStream.read(dataBuffer, 0, math.min(remaining, UPLOAD_SIZE))
+        outputStream.write(dataBuffer, 0, read)
+        remaining -= read
+      }
+    } catch {
+      case e: IOException => logError("Exception occurred while writing some profiler output: ", e)
+      case e@(_: IllegalArgumentException | _: IllegalStateException) =>
+        logError("Some profiler output not written." +
+          " Exception occurred in profiler native code: ", e)
+      case e: Exception => logError("Some profiler output not written. Unexpected exception: ", e)
```

Comment:
I'm wondering what we can get in this case because the flag `writing` is still `true`. Do we need to keep `writing` because we need to invoke `finishWriting`? However, it seems that we cannot invoke `inputStream.close()` and `outputStream.close()` eventually.

Reply:
Got it. Since we have the error log, it might be okay for now.